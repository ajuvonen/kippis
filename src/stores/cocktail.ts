import {
  getByIngredients,
  getBySearchString,
  getDetails,
  getNonAlcoholic,
  getByFirstLetter,
} from '@/api';
import {SEARCHABLE_ALCOHOLS} from '@/utils/constants';
import type {FullDetailsCocktail, SearchResultCocktail, SearchStringProps} from '@/utils/types';
import {defineStore} from 'pinia';
import {compose, flatten, map, prop, sortBy, uniq, without} from 'ramda';

export const useCocktailStore = defineStore('cocktail', {
  state: () => ({
    currentSearch: {} as SearchStringProps,
    cocktailDetails: [] as FullDetailsCocktail[],
    preventFetch: false,
    searchResults: [] as SearchResultCocktail[],
    selection: new Set<number>(),
    highlightedCocktail: null as number | null,
  }),
  getters: {
    getCocktailDetails: (state) => (id: number) => state.cocktailDetails.find((cocktail) => cocktail.id === id) || null,
    // Get a "shopping list" of ingredients. They should be unique, sorted and without some obvious items
    getAllIngredients(state): string[] {
      return compose(
        sortBy(prop(0)),
        without(['water', 'ice']),
        uniq,
        flatten<string[][]>,
        map((id: number) =>
          // Make cream, egg, and citrus fruits show up as their base form 
          this.getCocktailDetails(id)!.ingredients.map(({ingredient}) =>
            ingredient.replace(/^whipped | peel$| spiral$| white$| yolk$/, ''),
          ),
        ),
      )(Array.from(state.selection));
    },
  },
  actions: {
    async addToSelection(id: number) {
      await this.fetchCocktail(id);
      this.selection.add(id);
    },
    async fetchCocktail(id: number) {
      if (!this.cocktailDetails.some((cocktail) => cocktail.id === id)) {
        this.cocktailDetails.push(await getDetails(id));
      }
    },
    async openCocktailModal(id: number) {
      await this.fetchCocktail(id);
      this.highlightedCocktail = id;
    },
    removeFromSelection(id: number) {
      this.selection.delete(id);
    },
    async search(searchString: string) {
      if (searchString.length === 1) {
        this.searchResults = await getByFirstLetter(searchString);
      } else {
        this.searchResults = await getBySearchString(searchString);
      }
    },
    async searchNonAlcoholic() {
      this.searchResults = await getNonAlcoholic();
    },
    async searchWithTag(tag: string) {
      this.searchResults = await getByIngredients(
        SEARCHABLE_ALCOHOLS.find((alcohol) => tag === alcohol.tag)?.ingredients!,
      );
    },
  },
});
