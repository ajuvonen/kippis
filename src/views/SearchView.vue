<script setup lang="ts">
import {ref, watch} from 'vue';
import {storeToRefs} from 'pinia';
import SearchField from '@/components/SearchField.vue';
import {useCocktailStore} from '@/stores/cocktail';
import {ALCOHOLS} from '@/utils/constants';

const props = defineProps<{
  searchString?: string;
  tag?: string;
}>();

const searchString = ref('');
const cocktailStore = useCocktailStore();
const {search, searchWithTag} = cocktailStore;
const {searchResults} = storeToRefs(cocktailStore);

watch(props, (value) => {
  if (value.tag && ALCOHOLS.filter(({tag}) => tag === value.tag)) {
    searchWithTag(value.tag);
  } else if (value.searchString) {
    searchString.value = value.searchString.trim().slice(0, 20);
    search(searchString.value);
  }
}, {immediate: true});

</script>
<template>
  <main>
    <search-field />
  </main>
</template>
@/utils/constants