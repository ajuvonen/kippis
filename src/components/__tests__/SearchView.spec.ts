import {describe, it, expect, beforeEach} from 'vitest';
import {mount} from '@vue/test-utils';
import SearchView from '@/views/SearchView.vue';
import {useCocktailStore} from '@/stores/cocktail';
import {testSearchResults} from './mswHandlers';

describe('SearchView', () => {
  let cocktailStore: ReturnType<typeof useCocktailStore>;

  beforeEach(() => {
    cocktailStore = useCocktailStore();
  });

  it('renders correctly', () => {
    const wrapper = mount(SearchView);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('searches with tag', () => {
    mount(SearchView, {
      props: {
        tag: 'gin',
      },
    });

    expect(cocktailStore.search).not.toHaveBeenCalled();
    expect(cocktailStore.searchNonAlcoholic).not.toHaveBeenCalled();
    expect(cocktailStore.searchWithTag).toHaveBeenCalledWith('gin');
  });

  it('searches non-alcoholic', () => {
    mount(SearchView, {
      props: {
        tag: 'virgin',
      },
    });

    expect(cocktailStore.search).not.toHaveBeenCalled();
    expect(cocktailStore.searchWithTag).not.toHaveBeenCalled();
    expect(cocktailStore.searchNonAlcoholic).toHaveBeenCalled();
  });

  it('searches with search string', () => {
    mount(SearchView, {
      props: {
        searchString: 'tom',
      },
    });

    expect(cocktailStore.searchNonAlcoholic).not.toHaveBeenCalled();
    expect(cocktailStore.searchWithTag).not.toHaveBeenCalled();
    expect(cocktailStore.search).toHaveBeenCalledWith('tom');
  });

  it('does not search without search string', () => {
    mount(SearchView, {
      props: {
        searchString: '',
      },
    });

    expect(cocktailStore.searchNonAlcoholic).not.toHaveBeenCalled();
    expect(cocktailStore.searchWithTag).not.toHaveBeenCalled();
    expect(cocktailStore.search).not.toHaveBeenCalled();
  });

  it('cuts search string', () => {
    mount(SearchView, {
      props: {
        searchString: 'quite a long string with spaces',
      },
    });

    expect(cocktailStore.searchNonAlcoholic).not.toHaveBeenCalled();
    expect(cocktailStore.searchWithTag).not.toHaveBeenCalled();
    expect(cocktailStore.search).toHaveBeenCalledWith('quite a long string ');
  });

  it('prevents fetch', () => {
    cocktailStore.preventFetch = true;
    mount(SearchView, {
      props: {
        searchString: 'test',
      },
    });

    expect(cocktailStore.searchNonAlcoholic).not.toHaveBeenCalled();
    expect(cocktailStore.searchWithTag).not.toHaveBeenCalled();
    expect(cocktailStore.search).not.toHaveBeenCalled();
    expect(cocktailStore.preventFetch).toBe(false);
  });

  it('displays result count', () => {
    cocktailStore.preventFetch = true;
    cocktailStore.searchResults = [testSearchResults[0][1], testSearchResults[1][1]];
    const wrapper = mount(SearchView, {
      props: {
        searchString: 'test',
      },
    });

    expect(wrapper.find('h1').text()).toBe('Results (2)');
  });
});
