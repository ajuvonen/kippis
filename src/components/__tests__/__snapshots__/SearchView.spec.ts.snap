// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchView > renders correctly 1`] = `
"<transition-stub data-v-418939cb=\\"\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
  <!--v-if-->
</transition-stub>
<main class=\\"flex-1\\"><label data-v-b87fbe66=\\"\\" for=\\"search-box\\" class=\\"block text-center text-3xl mt-4 mb-4\\">Search for Recipes</label>
  <div data-v-b87fbe66=\\"\\" class=\\"flex justify-center divide-x mb-4\\"><input data-v-b87fbe66=\\"\\" id=\\"search-box\\" type=\\"search\\" maxlength=\\"20\\" class=\\"rounded-tr-none rounded-br-none pb-1\\"><button data-v-b87fbe66=\\"\\" class=\\"rounded-tl-none rounded-bl-none\\"><svg data-v-b87fbe66=\\"\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" class=\\"w-5 h-5\\">
        <path d=\\"M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z\\" class=\\"transition-all ease-in-out\\"></path>
      </svg><span data-v-b87fbe66=\\"\\">Search</span></button></div>
  <ul data-v-b87fbe66=\\"\\" class=\\"flex flex-wrap justify-center\\" aria-label=\\"Recipe Categories\\">
    <li data-v-b87fbe66=\\"\\" class=\\"inline\\"><a data-v-b87fbe66=\\"\\" href=\\"/search?tag=gin\\" class=\\"flex items-center link-button\\">Gin</a></li>
    <li data-v-b87fbe66=\\"\\" class=\\"inline\\"><a data-v-b87fbe66=\\"\\" href=\\"/search?tag=rum\\" class=\\"flex items-center link-button\\">Rum</a></li>
    <li data-v-b87fbe66=\\"\\" class=\\"inline\\"><a data-v-b87fbe66=\\"\\" href=\\"/search?tag=vodka\\" class=\\"flex items-center link-button\\">Vodka</a></li>
    <li data-v-b87fbe66=\\"\\" class=\\"inline\\"><a data-v-b87fbe66=\\"\\" href=\\"/search?tag=whiskey\\" class=\\"flex items-center link-button\\">Whiskey</a></li>
    <li data-v-b87fbe66=\\"\\" class=\\"inline\\"><a data-v-b87fbe66=\\"\\" href=\\"/search?tag=virgin\\" class=\\"flex items-center link-button\\">Virgin</a></li>
    <li data-v-b87fbe66=\\"\\"><button data-v-b87fbe66=\\"\\"><svg data-v-b87fbe66=\\"\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" class=\\"w-5 h-5\\">
          <path d=\\"M12,11.18C15.3,8.18 17,6.64 17,4.69C17,3.19 15.75,2 14.25,2C13.39,2 12.57,2.36 12,3C11.43,2.36 10.61,2 9.69,2C8.19,2 7,3.25 7,4.75C7,6.64 8.7,8.18 12,11.18M11.18,12C8.18,8.7 6.64,7 4.69,7C3.19,7 2,8.25 2,9.75C2,10.61 2.36,11.43 3,12C2.36,12.57 2,13.39 2,14.31C2,15.81 3.25,17 4.75,17C6.64,17 8.18,15.3 11.18,12M12.83,12C15.82,15.3 17.36,17 19.31,17C20.81,17 22,15.75 22,14.25C22,13.39 21.64,12.57 21,12C21.64,11.43 22,10.61 22,9.69C22,8.19 20.75,7 19.25,7C17.36,7 15.82,8.7 12.83,12M12,12.82C8.7,15.82 7,17.36 7,19.31C7,20.81 8.25,22 9.75,22C10.61,22 11.43,21.64 12,21C12.57,21.64 13.39,22 14.31,22C15.81,22 17,20.75 17,19.25C17,17.36 15.3,15.82 12,12.82Z\\" class=\\"transition-all ease-in-out\\"></path>
        </svg><span data-v-b87fbe66=\\"\\">I'm Feeling Lucky</span></button></li>
  </ul>
  <h1>Results (0)</h1>
  <ul class=\\"flex flex-auto flex-wrap gap-6 content-start justify-center\\"></ul>
</main>"
`;
